report_qweb_encrypt/models/ir_actions_report.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2020 Creu Blanca
+# Copyright 2020 Ecosoft Co., Ltd.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from io import BytesIO
+
+from PyPDF2 import PdfFileReader, PdfFileWriter
+
+from odoo import _, fields, models
+from odoo.exceptions import ValidationError
+from odoo.tools.safe_eval import safe_eval
+
+
+class IrActionsReport(models.Model):
+    _inherit = "ir.actions.report"
+
+    encrypt = fields.Selection(
+        [("manual", "Manual Input Password"), ("auto", "Auto Generated Password")],
+        string="Encryption",
+        help="* Manual Input Password: allow user to key in password on the fly. "
+        "This option available only on document print action.\n"
+        "* Auto Generated Password: system will auto encrypt password when PDF "
+        "created, based on provided python syntax.",
+    )
+    encrypt_password = fields.Char(
+        help="Python code syntax to gnerate password.",
+    )
+
+    def _render_qweb_pdf(self, reportname, res_ids=None, data=None):
+        document, ttype = super()._render_qweb_pdf(
+            reportname, res_ids=res_ids, data=data
+        )
+        if res_ids:
+            encrypt_password = self._context.get("encrypt_password")
+            report = self._get_report_from_name(reportname).with_context(
+                encrypt_password=encrypt_password
+            )
+            password = report._get_pdf_password(res_ids[:1])
+            document = self._encrypt_pdf(document, password)
+        return document, ttype
+
+    def _get_pdf_password(self, res_ids):
+        encrypt_password = False
+        if self.encrypt == "manual":
+            encrypt_password = self._context.get("encrypt_password", False)
```

Comment:
You're removing comments above this line

Reply:
I originally deleted them because I believe they were no longer relevant, but they still are, so I've added the back, thanks